refactor: A value of empty string should apply the default value

darinspivey opened the pull request:
**fix: Correct boolean test to not use `required`**

Since required has special handling now in that it will
not allow blank inputs, this test should just see if
the validation passes for a bare-bones definition. Check
the resulting value to make sure it's `false`

---
**feat: allowEmpty() will allow '' to be a valid value**

In certain cases, we may want to have a default for use if
the env var is *omitted*, but allow the value of the empty
string to be a valid value (without triggering the default
value). This allows for that.

`.string('db-options).default('autoConnect=true').allowEmpty()`

would allow the db-options to be set to an empty string and
NOT auto connect in this example.

---

**style: Give custom errors their own files**

Some errors were inline and others were files. This cleans
up the main code and gets everything consistent into their
own error class files.

---

**package: CI test with newer versions of node**


---

**deps: eslint-config-logdna@6.1.0**


---

**deps: tap@16.2.0**


---

**refactor: A value of empty string should apply the default value**

When a value of '' is given, this should trigger the default
value to be used. A common use case is in k8s manifests where
mustache templates may blindly set an env var to a template
value. If that value is empty, the empty string will be used
and can cause the code to throw for types like `enum`. This
change allows us to skip setting defaults in such manifests
and let `Config` handle the defaults.

Also, the `_validateString` function was removed due to it
not being covered. The condition for `_required` is covered
in `_populate`, leaving the string validation function to
only be used for lists. This error check was simply moved
to the list validation.

In addition, a new mutex error has been added for using
`required` and `default` together, which doesn't make much
sense at a glance.

BREAKING CHANGE: This change affects the way that default values
are assigned, and the default value will now be used if the
env var's value is ''. Also, using `.required()` and `.default()`
in the same definition will now result in a mutex error.

darinspivey committed:
refactor: A value of empty string should apply the default value

When a value of '' is given, this should trigger the default
value to be used. A common use case is in k8s manifests where
mustache templates may blindly set an env var to a template
value. If that value is empty, the empty string will be used
and can cause the code to throw for types like `enum`. This
change allows us to skip setting defaults in such manifests
and let `Config` handle the defaults.

Also, the `_validateString` function was removed due to it
not being covered. The condition for `_required` is covered
in `_populate`, leaving the string validation function to
only be used for lists. This error check was simply moved
to the list validation.

In addition, a new mutex error has been added for using
`required` and `default` together, which doesn't make much
sense at a glance.

BREAKING CHANGE: This change affects the way that default values
are assigned, and the default value will now be used if the
env var's value is ''. Also, using `.required()` and `.default()`
in the same definition will now result in a mutex error.

darinspivey committed:
deps: tap@16.2.0

darinspivey committed:
deps: eslint-config-logdna@6.1.0

darinspivey committed:
package: CI test with newer versions of node

This also removes older versions of node, 10 and 12.

jsumners commented on Jenkinsfile:
```diff
@@ -36,7 +36,7 @@ pipeline {
           axis {
             name 'NODE_VERSION'
-            values '10', '12', '14'
+            values '10', '12', '14', '16', '17'
```
Are we able to drop 10 and 12?

darinspivey replied:
we could, but it's public, so technically we still support older versions.

jsumners commented on lib/definition.js:
```diff
@@ -354,18 +343,18 @@ module.exports = class Definition {
     assert(this._list_type, `Invalid Env Config (${this._name}): Missing .type()`)
-    const is_array = Array.isArray(this._value)
-    if (this._required && (!is_array || !this._value.length)) {
+    if (!Array.isArray(this._value)) return
```
nit: I find coercive checks like `!foo` to be less clear than an explicit boolean check:

```suggestion
    if (Array.isArray(this._value) === false) return
```

darinspivey replied:
Yeah, styles are tough.  Personal preference and such.  We typically don't do that style (for no particular reason other than maybe it's more typing?), and I don't have a strong preference on this one.

I could change it here, but it wouldn't be consistent with the other conditionals, and I think for style, that "rule" trumps all--try to follow the style of the overall code.

jsumners reviewed (commented):
Is it common to explicitly set environment variables to an empty string? Or does the Mustache system set an empty string when it cannot find a referenced environment variable?

esatterwhite commented on lib/definition.js:
```diff
@@ -354,18 +343,18 @@ module.exports = class Definition {
+    if (this._required && !this._value.length) {
       throw new MissingEnvError({
-        type: 'list'
+        type: this._type
```
I'm not sure this is accurate. its a list of type

darinspivey replied:
Are you saying to hardcode the type to `list` again? I just pasted this code, so that'd be fine.

esatterwhite replied:
I guess I'm asking. Does the error still say something that is helpful / accurate ? I'm not sure what the type value is when it's a list

esatterwhite replied:
I think it's kind of confusing either way. The error should probably say `[]${type}`

darinspivey replied:
Gonna leave this as-is for now.  This is getting large and I need a break :]

darinspivey commented:
I discovered in testing that you can do `.required().default('x')`, which doesn't make much sense.  Something that is `required` should not need a default.  I think while this is a breaking change, we should throw a commit on here that makes `required` and `default` mutually exclusive.

mdeltito commented:
With this change, it seems like it will be impossible to override e.g. a default value of `'foo'` with `''`. Should this be only for optional config?

esatterwhite commented on lib/definition.js:
```diff
@@ -354,18 +343,18 @@ module.exports = class Definition {
-
     const type = this._list_type.charAt(0).toUpperCase() + this._list_type.slice(1)
     const method = `_validate${type}`
+    if (!this[method]) return
```
What doesn't have a validation method?

darinspivey replied:
string.  The validation got a little redundant and the string validation didn't appear to be doing much except in the list context.  Here, it does `String(value).split`, so you've always got an array of strings, or an empty array.  I removed this function because it was impossible to cover since `_populate` handles the `required` checks

darinspivey committed:
style: Give custom errors their own files

Some errors were inline and others were files. This cleans
up the main code and gets everything consistent into their
own error class files.

darinspivey committed:
feat: allowEmpty() will allow '' to be a valid value

In certain cases, we may want to have a default for use if
the env var is *omitted*, but allow the value of the empty
string to be a valid value (without triggering the default
value). This allows for that.

`.string('db-options).default('autoConnect=true').allowEmpty()`

would allow the db-options to be set to an empty string and
NOT auto connect in this example.

jsumners commented on README.md:
```diff
@@ -171,14 +171,16 @@ or do not satisfy the requirements, an error will be thrown.
 
-Specifies that this env var is required. By default, env vars are optional.
+Specifies that this env var is required to have a value (env vars are optional by default). This means
+that a value must be provided, and it **cannot** be the empty string, `''`.
```
Is this a change in behavior for `.required()`?

darinspivey replied:
yes, it is.  Most of us agreed that if `required`, you shouldn't be able to give a falsey empty string.  Lemme know if there's a use case that breaks this thought.  It's a major, so folks might have to change a few things to upgrade.

jsumners replied:
The PR is labeled as "fix". So I assume the CI/CD tooling will bump `z` in `x.y.z`. The PR should be adjusted so that it is a new major. With that done, I'd argue dropping Node 10 and 12 is viable.

darinspivey replied:
The title is old.  I updated the commit to be `refactor` with `BREAKING CHANGE`.

We can drop the old nodes, sure.

darinspivey replied:
I removed the old versions of node, and I also switch Jenkins to use our `tools` to load node and test/release.  Not using docker is faster, and these 2 changes together increased speed by 50% or so.  Thanks, @jsumners

mdeltito commented on README.md:
```diff
@@ -204,10 +206,28 @@ Returns `this` to allow chaining.
 
+#### `Definition#allowEmpty()`
+
+This tells the definition to allow the empty string, `''`, as a legitimate value for the
+env var. This empty value will be type casted appropriately for the definition's `type`.
+A common use for this is to allow `''` as a value for a `.string()` definition.
```
This is a total nit, but I think it'd be clearer to instead talk about "empty value" for the type, rather than mention empty strings. I think it makes sense when you really think about it (it's env vars, they are always strings) but the mental-model of this package is around the _types_.

Just a suggestion:

> `allowEmpty()` tells the definition to accept an empty value in place of this provided non-empty `.default()`,
for cases where an empty value is valid and/or expected. Without this option, the default value will be used when
an empty value is detected for a given type. A common use for this is to allow `''` as a value for a 
`.string()` definition.

> The empty values for the config types supported by this package are as follows:

darinspivey replied:
That's fine.  It always made sense to me to think about strings because the empty string is a prerequisite to ending up with whatever "empty" type.  I'll change this, np.

darinspivey replied:
done.

mdeltito commented on lib/boolean-values.js:
```diff
@@ -20,7 +20,6 @@ const FALSE_VALUES = new Set([
 , 'no'
 , 'N'
 , 'n'
-, ''
```
do we actually want to remove this? I think empty string should still evaluate to false for a boolean - I'm assuming this got tangled up in the default values?

darinspivey replied:
easier said that done.  Since the empty string triggers special features now, this means something.

I can have another perhaps.  I don't disagree that it's intuitive that `''` means false.  I had taken it out early on cause it broke a bunch of things.

darinspivey replied:
ok, I remember now.  This one gets a little tricky.  There's a test that iterates through all of these "false" values and makes sure it passes validation.  However, the definition it uses has `required()`, and by our new rules, `''` is not a valid value for required.

SO--if we want this to be the case only for boolean (that '' is actually a valid `false`), it'll have to have special handling.  OR--we could not mutate inputs, and say that if something is `required`, then you have to provide `'false'` 🤷🏻‍♂️

darinspivey replied:
nvm, I just fixed the test.  It shouldn't be using `required`

mdeltito replied:
ahhh that makes sense, nice 👍

mdeltito commented on lib/definition.js:
```diff
@@ -197,38 +155,50 @@ module.exports = class Definition {
+    }
+
+    if (value_is_missing) {
+      if (this._default !== null && !this._allow_empty) {
```
is it worth mentioning somewhere that `allowEmpty` only has effect when there is a `default()`? I could see that being confusing - empty values are always "allowed", this just controls if you can override a default with empty.

darinspivey replied:
Probably should mention that, yeah.  After all, you don't need `allowEmpty` if there's no default.

darinspivey replied:
done.  Actually, your suggestion kind of says it all, but I added a TLDR sentence to be as clear as possible.

mdeltito reviewed (commented):
thanks for working through this 😄

ligerzero459 reviewed (dismissed):
Nice job, no additional feedback from what Mike asked about. Feel free to overwrite the approval as needed

darinspivey committed:
fix: Correct boolean test to not use `required`

Since required has special handling now in that it will
not allow blank inputs, this test should just see if
the validation passes for a bare-bones definition. Check
the resulting value to make sure it's `false`

evanlucas commented on Jenkinsfile:
```diff
@@ -36,16 +36,21 @@ pipeline {
         }
 
+        tools {
+          nodejs "NodeJS ${NODE_VERSION}"
```
I’d do this in docker still so we get the latest patch version for each

darinspivey replied:
gah

darinspivey replied:
What's the risk of not being at the current patch version.  As long as the ones we're running are stable, that should be good enough since we'll never be able to always be current.  I feel docker is unnecessary overhead for a standalone package like this, no?

darinspivey replied:
We should be patching Jenkins then too--to use docker is a workaround for Jenkins not being current, which feels like the thing to fix :]

darinspivey replied:
The commit has been dropped to keep this moving along.
